src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```
@@ -0,0 +1,353 @@
+
+                // +----+--------+
+                // |VER | METHOD |
+                // +----+--------+
+                // | 1  |   1    |
+                // +----+--------+
+                await ReadToFillAsync(stream, buffer.AsMemory(0, 2), async).ConfigureAwait(false);
+                VerifyProtocolVersion(ProtocolVersion5, buffer[0]);
+
+                switch (buffer[1])
+                {
+                    case METHOD_NO_AUTH:
+                        // continue
+                        break;
+
+                    case METHOD_USERNAME_PASSWORD:
+                        {
+                            // https://tools.ietf.org/html/rfc1929
+                            if (credentials is null)
+                            {
+                                // If the server is behaving well, it shouldn't pick username and password auth
+                                // because we don't claim to support it when we don't have credentials.
+                                // Just being defensive here.
+                                throw new SocksException(SR.net_socks_auth_required);
+                            }
+
+                            // +----+------+----------+------+----------+
+                            // |VER | ULEN |  UNAME   | PLEN |  PASSWD  |
+                            // +----+------+----------+------+----------+
+                            // | 1  |  1   | 1 to 255 |  1   | 1 to 255 |
+                            // +----+------+----------+------+----------+
+                            buffer[0] = SubnegotiationVersion;
+                            byte usernameLength = checked((byte)Encoding.UTF8.GetBytes(credentials.UserName, buffer.AsSpan(2)));
+                            buffer[1] = usernameLength;
+                            byte passwordLength = checked((byte)Encoding.UTF8.GetBytes(credentials.Password, buffer.AsSpan(3 + usernameLength)));
```

We should throw a meaningful exception if the username/password are too long.